### PR TITLE
Keycloak Integration

### LINKED ISSUE #340: Authentication using keycloak
- [x] yaptide/yaptide#475
- [x] yaptide/ui#1128
- [x] yaptide/yaptide#342
- [x] yaptide/yaptide#455
- [x] yaptide/yaptide#343

### LINKED ISSUE #342: adjust user model to handle key cloak data

### LINKED ISSUE #343: having keycloak token, retrieve SSH credentials from some endpoint

### PR DESCRIPTION
**Pull request by p1003**
PR following the changes being introduced in https://github.com/yaptide/ui/pull/1086.

TODO list:
- [x]  redefinition of UserModel in database,
- [x]  redefinition of ClusterModel in database,
- [x]  storing certs in database,
- [x]  update of connection with clusters.

### CODE REVIEW

File: Dockerfile-flask [hunk 1]
```diff
@@ -16,4 +17,4 @@ RUN pip install --no-cache-dir -r requirements.txt
 COPY yaptide ./yaptide
 
 # Run Flask app
-ENTRYPOINT ["flask", "run", "--host=0.0.0.0", "--port=6000"]
+ENTRYPOINT ["flask", "run", "--host=0.0.0.0", "--port=6000", "--debug"]
```
Comment (grzanka) [inline Dockerfile-flask]: is thi debug needed ?

File: Dockerfile-flask [hunk 2]
```diff
@@ -6,6 +6,7 @@ EXPOSE 6000
 # set environment variables
 ENV FLASK_APP=yaptide.application
 ENV FLASK_SQLALCHEMY_DATABASE_URI=sqlite:////usr/local/app/yaptide/data/main.db
+ENV CERT_AUTH_URL=https://ccm-dev.kdm.cyfronet.pl/key
```
Comment (grzanka) [inline Dockerfile-flask]: move to https://github.com/yaptide/deploy/blob/main/plgrid_vars.yml

Comment (grzanka) [reply Dockerfile-flask]: here this may be set to empty value or not set at all

File: yaptide/admin/db_manage.py [hunk 3]
```diff
@@ -55,7 +57,10 @@ def list_users(**kwargs):
     if con is None or metadata is None or engine is None:
         return None
     users = db.Table(TableTypes.USER.value, metadata, autoload=True, autoload_with=engine)
-    query = db.select([users])
+    yaptide_users = db.Table(TableTypes.YAPTIDEUSER.value, metadata, autoload=True, autoload_with=engine)
+    joined_users = users.join(yaptide_users, users.c.id == yaptide_users.c.id)
```
Comment (grzanka) [inline yaptide/admin/db_manage.py]: What do you mean by joined users ? Add a comment explaining this code. Do you mean local users and keycloak users ? 
Why do you make join based on having the same id ?

File: yaptide/batch/batch_methods.py [hunk 4]
```diff
@@ -18,27 +18,37 @@
...
+        }
     )
+    return con
+
+
+def submit_job(payload_dict: dict, files_dict: dict, user: KeycloakUserModel, cluster: ClusterModel) -> dict:
+    """Dummy version of submit_job"""
+    utc_time = int(datetime.utcnow().timestamp()*1e6)
+
+    con = get_connection(user=user, cluster=cluster)
```
Comment (grzanka) [inline yaptide/batch/batch_methods.py:45]: At which point we connect to the cluster ? At getting connection or at calling `con.run` ?

Which will fail if we have invalid certificate ?

Comment (grzanka) [reply yaptide/batch/batch_methods.py:45]: You should handle here the case where cert is invalid

File: yaptide/admin/db_manage.py (hunk 3 above)
Comment (p1003) [reply yaptide/admin/db_manage.py]: Outdated

File: Dockerfile-flask (hunk 1 above)
Comment (grzanka) [inline Dockerfile-flask]:
```suggestion
ENTRYPOINT ["flask", "run", "--host=0.0.0.0", "--port=6000"]
```

File: Dockerfile-flask (hunk 2 above)
Comment (grzanka) [inline Dockerfile-flask]:
```suggestion
ENV CERT_AUTH_URL=""
```

**Review by grzanka — APPROVED** [review]

**Pull request merged**